File: Clojure/problem044.clj
#! /usr/bin/env clojure
; Problem 44 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=44
(defn penta? [y]
  (let [d (+ 1 (* 24 y))
        d_sq (int (Math/sqrt d))]
    (and (= (* d_sq d_sq) d)
         (zero? (mod (inc d_sq) 6)))))

(defn penta [n]
  (/ (* n (dec (* 3 n))) 2))

(loop [k 2 j 1]
  (let [pk (penta k) pj (penta j)]
    (if (and (penta? (+ pk pj))
             (penta? (- pk pj)))
      (println pk pj ", D =" (- pk pj))

      (if (< j (dec k))
        (recur k (inc j))
        (recur (inc k) 1)))))

File: Kotlin/problem044.kt
// Problem 44 - Project Euler
// http://projecteuler.net/index.php?section=problems&id=44
import kotlin.math.*

fun is_penta(i: Int): Boolean {
    val d = 1 + 24 * i
    val d_sq = sqrt(d.toDouble()).toInt()
    return d_sq * d_sq == d && (d_sq + 1) % 6 == 0
}

fun penta(i: Int): Int {
    return i * (3 * i - 1) / 2
}

fun main() {
    var k = 2
    loop@ while (true) {
        val pk = penta(k)

        for (j in 1..k) {
            val pj = penta(j)
            val p1 = pk - pj
            val p2 = pk + pj

            if (is_penta(p1) && is_penta(p2)) {
                println(pk)
                println(pj)
                println(p1)
                break@loop
            }
        }

        k += 1
    }
}
